# Exploring fundamentals of datetime

In [2]:
from datetime import datetime
from datetime import date

## Datetime
Datetime provides access to all components of time (year, month, day, hour, second, millisecond)

In [46]:
# create today's date
datetime.today()

datetime.datetime(2018, 6, 27, 11, 0, 44, 600286)

In [47]:
# let's see the type of the date
today = datetime.today() 
type(today)

datetime.datetime

You can access specific components of the datetime using accessor methods

In [52]:
# year
today.year

2018

In [53]:
# month
today.month

6

In [54]:
# day
today.day

27

In [55]:
# hour
today.hour

11

In [56]:
# minute
today.minute

0

In [57]:
# second
today.second

45

In [58]:
# microsecond
today.microsecond

30906

In [60]:
# timezone
today.tzinfo

## Date
Date provides access to date components only - no time information (day, month, year)

In [10]:
# show today
date.today()

datetime.date(2018, 6, 27)

In [11]:
# see the type of date variable 
today = date.today()
type(today)

datetime.date

You can access specific components of the date using accessor methods

In [61]:
# year
today.year

2018

In [62]:
# month
today.month

6

In [63]:
# day
today.day

27

When we subract/add dates, we produce timedelta objects

In [22]:
# declaring a date for christmas
christmas = date(2018, 12, 25)

# subtract christmas from today
christmas - today

datetime.timedelta(181)

We can see we get a timedelta of 181 days.

In [23]:
## access the days component of the delta
(christmas - today).days

181

In [25]:
# how can we compare dates?
if christmas is not today:
    print('Sorry, there is still ' + str((christmas-today).days) + ' until christmas.')
else:
    print('Its Christmas')

Sorry, there is still 181 until christmas.


## Timedelta
Timedeltas allow us to represent spans of time within python

In [18]:
from datetime import timedelta

In [69]:
# create a timedelta
t = timedelta(days=4, hours=10)
t

datetime.timedelta(4, 36000)

In [68]:
type(t)

datetime.timedelta

A couple items to note
* timedelta does not have hours element - it will create an error
* timedelta only stores days, seconds, microseconds
* seconds displays the number of seconds in the 'hours' component of the timedelta
* seconds can only show a maximum of 1 day (because we can only have 24 hours max - before the value turns into a 'day' element

In [64]:
# days
t.days

4

In [65]:
# seconds
t.seconds

36000

In [66]:
# microseconds
t.microseconds

0

If you want to show number of hours - you can convert seconds into hours

In [32]:
# 60 seconds in a minute, 60 minutes in an hour
t.seconds / 60 / 60

10.0

In [33]:
# 3600 seconds in an hour
t.seconds / 3600

10.0

We can also add timedeltas to dates or datetimes

In [70]:
today = datetime.today()
today

datetime.datetime(2018, 6, 27, 12, 19, 46, 525601)

In [73]:
eta = timedelta(hours=12)
eta

datetime.timedelta(0, 43200)

In [74]:
today+eta

datetime.datetime(2018, 6, 28, 0, 19, 46, 525601)

In [40]:
print(str(today+eta))

2018-06-27 16:52:19.519928
